pkg/util/pvc/pvc_compare_utils.go

```diff
@@ -0,0 +1,108 @@
+package pvc
+
+import v1 "k8s.io/api/core/v1"
+
+type CheckClaimFn = func(*v1.PersistentVolumeClaim, *v1.PersistentVolumeClaim) bool
+
+// CheckClaimCompatibleWithoutSize PersistentVolumeClaim Compare function
+func CheckClaimCompatibleWithoutSize(claim, template *v1.PersistentVolumeClaim) bool {
+	// when there is default sc,
+	// template StorageClassName is nil but claim is not nil
+	if template.Spec.StorageClassName != nil &&
+		claim.Spec.StorageClassName != nil &&
+		*claim.Spec.StorageClassName != *template.Spec.StorageClassName {
+		return false
+	}
+	// use set compare?
+	if len(claim.Spec.AccessModes) != len(template.Spec.AccessModes) {
+		return false
+	}
+	for i, mode := range claim.Spec.AccessModes {
+		if template.Spec.AccessModes[i] != mode {
+			return false
+		}
+	}
+	return true
+}
+
+// CheckPatchPVCCompleted PersistentVolumeClaim Compare function
+func CheckPatchPVCCompleted(claim, template *v1.PersistentVolumeClaim) bool {
+	compatible, ready := PVCCompatibleAndReady(claim, template)
+	if compatible && ready {
+		return true
+	}
+	if compatible {
+		pending := false
+		for _, condition := range claim.Status.Conditions {
+			if condition.Type == v1.PersistentVolumeClaimFileSystemResizePending {
+				pending = true
+			}
+		}
+		// if pending, patch PVC completed
+		return pending
+	}
+	return false
+}
+
+func CompareWithCheckFn(claim, template *v1.PersistentVolumeClaim, cmp CheckClaimFn) (matched, cmpResult bool) {
+	if !CheckClaimCompatibleWithoutSize(claim, template) {
+		return false, false
+	}
+	if cmp(claim, template) {
+		return false, true
+	}
+	return true, false
+}
+
+func PVCCompatibleAndReady(claim, template *v1.PersistentVolumeClaim) (compatible bool, ready bool) {
+	if !CheckClaimCompatibleWithoutSize(claim, template) {
+		return false, false
+	}
+	compatible = func(claim, template *v1.PersistentVolumeClaim) bool {
+		// claim >= template => compatible
+		// claim < template => not compatible (need patch by controller)
+		if claim.Spec.Resources.Requests.Storage() != nil &&
+			template.Spec.Resources.Requests.Storage() != nil &&
+			claim.Spec.Resources.Requests.Storage().Cmp(*template.Spec.Resources.Requests.Storage()) >= 0 {
+			return true
+		}
+		return false
+	}(claim, template)
+
+	ready = func(claim, template *v1.PersistentVolumeClaim) bool {
+		// cap >= spec => ready
+		// cap < spec => need storage expansion by csi
+		if claim.Status.Capacity.Storage() != nil &&
+			claim.Spec.Resources.Requests.Storage() != nil &&
+			claim.Status.Capacity.Storage().Cmp(*claim.Spec.Resources.Requests.Storage()) >= 0 {
+			return true
+		}
+		return false
```

Comment:
shall we check claim.status.allocatedResourceStatuses or claim.status.conditions ?